### ISSUE TITLE
Intermittent build failures with  "error CS0104: `X` is an ambiguous reference between `System.X` and `System.X`"

### ISSUE THREAD

**Issue by andrei-epure-sonarsource** [dotnet/roslyn#60259]
**Version Used**: 
- MSBuild version = "17.1.0+ae57d105c"
- dotnet sdk 6.0.200

**Steps to Reproduce**:

1. Clone https://github.com/duncanp-sonar/BugRepro_AmbiguousTypesNet6
2. Repro steps in the repo README _(TL;DR: do a sonar analysis when running the build, which injects sonar Roslyn analyzers; run the build multiple times until it fails; it intermittently fails - only ~20% of the runs)_

**Expected Behavior**: build & analysis is successfull

**Actual Behavior**:
Build fails intermittently with "ambiguous reference" errors e.g. 

>(CoreCompile target) -> d:\support\S4MSB\AmbiguousTypesNet6\Class2.cs(13,32): error CS0104: 'TimeSpan' is an ambiguous reference between 'System.TimeSpan' and 'System.TimeSpan' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj] d:\support\S4MSB\AmbiguousTypesNet6\Entities.cs(14,16): error CS0104: 'DateTime' is an ambiguous reference between 'System.DateTime' and 'System.DateTime' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj] d:\support\S4MSB\AmbiguousTypesNet6\Entities.cs(19,16): error CS0104: 'DateTime' is an ambiguous reference between 'System.DateTime' and 'System.DateTime' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj]

**Binary Logs**:

You can check the binary logs for a successful build and a failed one: https://github.com/duncanp-sonar/BugRepro_AmbiguousTypesNet6/tree/master/logs. 

Note: the `csc.exe` invocations are identical.

**Details**:

Initially reported on [sonar community](https://community.sonarsource.com/t/timespan-is-an-ambiguous-reference-between-system-timespan-and-system-timespan/59245/9).  Some users have reported other numbers for their projects (only 1 / 10 of the builds succeed). A user reported that by removing `ImplicitUsings`, the problem would stop manifesting itself.

Separately, this has been reported [on StackOverflow as well](https://stackoverflow.com/questions/71171933/error-cs0104-guid-is-an-ambiguous-reference-between-system-guid-and-system) (seems not to be related to the sonar Roslyn analyzers).

Can this be related to https://github.com/dotnet/roslyn/issues/59564 ?

Could you please help:
- Are you aware of non-deterministic behavior inside the compiler that could lead to the stochastic errors we are witnessing?
- Do you have any hints on what to investigate in order to drill down to the root cause of the issue? 

Thank you!

**Comment by RikkiGibson** [comment]
> * Are you aware of non-deterministic behavior inside the compiler that could lead to the stochastic errors we are witnessing?

I'm not aware of a specific issue that could be causing this. I think that #59564 is probably unrelated.

> * Do you have any hints on what to investigate in order to drill down to the root cause of the issue?

It would be useful to know if the failure still occurs when [`<ImplicitUsings>`](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#implicitusings) are disabled in the project.

**Comment by costin-zaharia-sonarsource** [comment]
We have now an easier way to reproduce this issue:

1. clone https://github.com/duncanp-sonar/BugRepro_AmbiguousTypesNet6
2. checkout `cz/stand-alone-repro` branch
3. run `.\repeat-build.ps1` script
4. In the last test I've made I had 1 build failure from 32 runs (a run is about 3seconds so it should be quick to reproduce)

> It would be useful to know if the failure still occurs when [<ImplicitUsings>](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#implicitusings) are disabled in the project.

I did a test and I could not reproduce the problem in 200 runs. I'm not sure if the issue is solved by this or if dramatically lowers the chance to happen :)

**Comment by RikkiGibson** [comment]
I haven't looked at this since I last interacted with the issue, but it does help when people chime in and let us know they are affected. I'll see if I can get this done by 17.3. Otherwise, if anyone in the community is interested in investigating, that could certainly help too. Because the csc invocations are identical here, I suspect an external/build system related root cause. For example, there might be a concurrency issue in the component which actually writes the global usings file out to disk. But it is also possible it is a compiler issue.

**Comment by fgroen** [comment]
This morning I had a feeling this was easy to reproduce in our environment. (About 80% build failures) 
Let me know what I can do to help you on this one.

**Comment by RikkiGibson** [comment]
It would be useful to know if the problem can be reproduced when using a compiler built in debug configuration. https://github.com/dotnet/roslyn/discussions/60840#discussioncomment-2595617

If it can be reproduced in debug configuration, we should be able to attach the debugger at the point where this diagnostic is issued and figure out more about what's going on.

https://github.com/dotnet/roslyn/blob/de252d8225d17e8e940764e441929cdd89770040/src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs#L1838-L1843

I will get to this issue eventually, but if you are interested in spending time to make progress along these steps, it may help get the issue resolved more quickly.

**Comment by fgroen** [comment]
Hmm, we're using azure devops for building and testing. I'll try if I can move this to a local install and when successful in reproducing, I can take a peek at getting the compiler in debug mode.

**Comment by jcouv** [comment]
FWIW, I wasn’t able to repro with a debug compiler, only a release compiler (used a local nuget package with a `FailFast` added, and referenced it). So it feels like it could be a race condition that's sensitive to certain timing.
The interesting thing from looking at the [dump file](https://1drv.ms/u/s!Ao4Mfha5mZFZjHun22qxciYWN2Pn?e=bFiRwe) is that the result set contains a literal duplicate (ie. same object ID/reference).

![image](https://user-images.githubusercontent.com/12466233/168691262-336fd1cb-a351-4738-8546-d293cac34d68.png)

**Comment by RikkiGibson** [comment]
I think this indicates that the compiler is at fault here. We'll want to investigate what might be different about lookup when global usings are in use. It could be that the same symbols are getting added multiple times to the same container concurrently, and when we look up in that container, we end up having "bogus" duplicates.

**Comment by AlekseyTs** [comment]
I think I figured out what could be the reason for the error. From the  [dump file](https://1drv.ms/u/s!Ao4Mfha5mZFZjHun22qxciYWN2Pn?e=bFiRwe), we can observe that ```global using System;``` and ```using System;``` are both in the list of usings to consider during name lookup. Both refer to the same namespace, but the namespace symbols are not identical, that is definitely going to keep them both in the list. Both namespaces are MergedNamespaceSymbols and those are supposed to be reference identical. I suspect the problem is in ```CachingDictionary``` which is used to back MergedNamespaceSymbol and create unique instances of child merged namespaces. It looks like it is failing to fulfill the last part in some multithreading scenarios. This can be demonstrated with the following unit-test that doesn't use multithreading, but carefully orders operations:
```
        [Fact]
        public void FullyPopulateRace()
        {
            CachingDictionary<int, FullyPopulateRaceHelper> lookup = null;
            FullyPopulateRaceHelper item = null;
            int itemAccessCount = 0;
            lookup = new CachingDictionary<int, FullyPopulateRaceHelper>(getElementsOfKey, getKeys, EqualityComparer<int>.Default);

            _ = lookup.Count;
            Assert.NotNull(item);
            Assert.Same(item, getItem());

            ImmutableArray<FullyPopulateRaceHelper> getElementsOfKey(int id)
            {
                itemAccessCount++;
                if (item is null && itemAccessCount == 1)
                {
                    item = getItem();
                }

                return ImmutableArray.Create<FullyPopulateRaceHelper>(new FullyPopulateRaceHelper());
            }

            HashSet<int> getKeys(IEqualityComparer<int> comparer)
            {
                return new HashSet<int>(new[] { 1 }, comparer);
            }

            FullyPopulateRaceHelper getItem()
            {
                return lookup[1][0];
            }
        }

        class FullyPopulateRaceHelper { }
```
 
Observed:
```
Microsoft.CodeAnalysis.UnitTests.CachingLookupTests.FullyPopulateRace [FAIL]
  Assert.Same() Failure
  Expected: FullyPopulateRaceHelper { }
  Actual:   FullyPopulateRaceHelper { }
  Stack Trace:
    src\Compilers\Core\CodeAnalysisTest\CachingLookupTests.cs(302,0): at Microsoft.CodeAnalysis.UnitTests.CachingLookupTests.FullyPopulateRace()
```

**Comment by cheesi** [comment]
@AlekseyTs Do you know, if this fix was included in the 17.3 release now?

**Comment by kirkpabk** [comment]
Disable unnecessary usings statements.
